packages/base64/src/decode.js

```diff
@@ -18,46 +18,55 @@ import { monodu64, padding } from './common.js';
  * messages.
  * @returns {Uint8Array} decoded bytes
  */
-const jsDecodeBase64 = (string, name = '<unknown>') => {
-  const data = new Uint8Array(Math.ceil((string.length * 4) / 3));
+export const jsDecodeBase64 = (string, name = '<unknown>') => {
+  // eslint-disable-next-line no-nested-ternary
+  const paddingLength = string.endsWith('==')
+    ? 2
+    : string.endsWith('=')
+    ? 1
+    : 0;
+  const data = new Uint8Array((string.length * 3) / 4 - paddingLength);
   let register = 0;
   let quantum = 0;
   let i = 0; // index in string
   let j = 0; // index in data
+  let done = false;
 
-  while (i < string.length && string[i] !== padding) {
-    const number = monodu64[string[i]];
+  for (const ch of string) {
+    if (done || ch === padding) {
+      done = true;
+      // Padding is only expected to complete a short chunk of two or three data characters
+      // (i.e., the last two in the `quantum` cycle of [0, 6, 12 - 8 = 4, 10 - 8 = 2]).
+      if (quantum === 4 || quantum === 2) {
+        quantum -= 2;
+        i += 1;
+        // eslint-disable-next-line no-continue
+        continue;
+      }
+      break;
+    }
+    const number = monodu64[ch];
     if (number === undefined) {
-      throw Error(`Invalid base64 character ${string[i]} in string ${name}`);
+      throw Error(`Invalid base64 character ${ch} in string ${name}`);
     }
     register = (register << 6) | number;
     quantum += 6;
     if (quantum >= 8) {
       quantum -= 8;
-      data[j] = register >>> quantum;
+      data[j] = register >> quantum;
       j += 1;
-      register &= (1 << quantum) - 1;
     }
     i += 1;
   }
 
-  while (i < string.length && quantum % 8 !== 0) {
-    if (string[i] !== padding) {
-      throw Error(`Missing padding at offset ${i} of string ${name}`);
-    }
-    i += 1;
-    quantum += 6;
-  }
-
   if (i < string.length) {
```

Comment:
Cache `string.length` at the beginning. But then in which case can there ever be garbage at the end?

Reply:
When a non-alphabet character is encountered, or anything after expected padding.

Reply:
Don't we throw if an invalid character is encountered?

One case where the number of characters iterated would differ from the string length is if any surrogate pairs are encountered, but again these would be invalid characters, and we'd have thrown already.

I maintain that this condition can never be true.

Reply:
```console
$ node --input-type=module -e '
  import "@endo/init";
  import { jsDecodeBase64 } from "./src/decode.js";
  console.log("valid", jsDecodeBase64("QQ=="));
  console.log("invalid");
  jsDecodeBase64("QQ===");
'
valid Uint8Array(1) [ 65 ]
invalid
(Error#1)
Error#1: Base64 string has trailing garbage = in string <unknown>

  at jsDecodeBase64 (packages/base64/src/decode.js:62:11)
  at packages/base64/[eval1]:6:3
```

Reply:
Oh right I missed the new break logic from the loop. I believe there's actually a bug allowing input like `QQ=*`

Reply:
Oh and it doesn't seem to check the register is empty once padding is reached

Reply:
It looks like [according to the RFC](https://datatracker.ietf.org/doc/html/rfc4648#section-3.5), a decoder "MAY" reject if the padding bits are not 0, but doesn't have to. I still believe this implementation will accept `QQ=*` for example, which is not compliant.

Reply:
> I believe there's actually a bug allowing input like `QQ=*`

Nice catch! Opened #1991 to add coverage and rebased this PR on top of its test changes.